### PR TITLE
Post Backend and Connection + UI Revamp

### PR DESCRIPTION
**Pull request by frinzekt**
### Backend
- Creation of routes, controllers and model for post with interaction to mongoDB (CRUD)

### Frontend
- Axios connection to the routes (Reading) 
- Revamping of postPreview, longPost

### Todo
- Creation + update (via next implementation of CKEditor)
- delete post

### Visual Changes
- Post Preview 
screencapture-localhost-3000-posts-1-2019-12-21-01_52_37

- Post
screencapture-localhost-3000-post-5dfd04a1f9858e7c11fc7889-2019-12-21-01_52_48

### Example of API used to produce post seen in the screenshot
```js
{
    "categories": [
        "categ1-1",
        "categ1-2"
    ],
    "comments": [
        "comment1",
        "comment2"
    ],
    "_id": "5dfd04a1f9858e7c11fc7889",
    "title": "Homelessness Awareness For 2019",
    "author": "author1",
    "previewDetails": "Consectetur est consequat incididunt anim ex quis pariatur commodo laborum laborum occaecat anim ea incididunt. Eiusmod ullamco est consequat ipsum aliqua ea qui irure aute commodo voluptate. Nisi incididunt fugiat exercitation et ea consectetur.Tempor minim minim cillum dolor reprehenderit adipisicing et eu veniam incididunt cupidatat duis. Consectetur velit duis anim sint ad commodo deserunt ex anim minim pariatur veniam tempor. Deserunt irure do ullamco amet consequat occaecat esse reprehenderit nulla sit ea. Sunt labore adipisicing commodo laboris sint ullamco.Ex nisi sint esse do nulla Lorem in tempor veniam ea. Id minim sint quis tempor ea in Lorem ullamco eu incididunt labore exercitation nostrud occaecat. Id Lorem ea eiusmod velit ad velit minim irure. Aliqua est mollit eu consectetur dolor cupidatat cupidatat elit.",
    "content": "<p>             Never in all their history have men been able truly to conceive of             the world as one: a single sphere, a globe, having the qualities of             a globe, a round earth in which all the directions eventually meet,             in which there is no center because every point, or none, is center             — an equal earth which all men occupy as equals. The airman&#39;s             earth, if free men make it, will be truly round: a globe in             practice, not in theory.           </p>           <p>             Science cuts two ways, of course; its products can be used for both             good and evil. But there&#39;s no turning back from science. The             early warnings about technological dangers also come from science.           </p>           <p>             What was most significant about the lunar voyage was not that man             set foot on the Moon but that they set eye on the earth.           </p>           <p>             A Chinese tale tells of some men sent to harm a young girl who, upon             seeing her beauty, become her protectors rather than her violators.             That&#39;s how I felt seeing the Earth for the first time. I could             not help but love and cherish her.           </p>           <p>             For those who have seen the Earth from space, and for the hundreds             and perhaps thousands more who will, the experience most certainly             changes your perspective. The things that we share in our world are             far more valuable than those which divide us.           </p>           <h2 class='section-heading'>Heading</h2>           <p>             There can be no thought of finishing for ‘aiming for the stars.’             Both figuratively and literally, it is a task to occupy the             generations. And no matter how much progress one makes, there is             always the thrill of just beginning.           </p>           <p>             There can be no thought of finishing for ‘aiming for the stars.’             Both figuratively and literally, it is a task to occupy the             generations. And no matter how much progress one makes, there is             always the thrill of just beginning.           </p>           <blockquote class='blockquote'>             <p class='mb-0'>               The dreams of yesterday are the hopes of today and the reality of               tomorrow. Science has not yet mastered prophecy. We predict too               much for the next year and yet far too little for the next ten.             </p>           </blockquote>           <p>             Spaceflights cannot be stopped. This is not the work of any one man             or even a group of men. It is a historical process which mankind is             carrying out in accordance with the natural laws of human             development.           </p>           <h2 class='section-heading'>Reaching for the Stars</h2>           <p>             As we got further and further away, it [the Earth] diminished in             size. Finally it shrank to the size of a marble, the most beautiful             you can imagine. That beautiful, warm, living object looked so             fragile, so delicate, that if you touched it with a finger it would             crumble and fall apart. Seeing this has to change a man.           </p>           <a href='#'>             <img class='img-fluid' src='post-sample-image.jpg' />           </a>           <span class='caption text-muted'>             To go places and do things that have never been done before – that’s             what living is all about.           </span>           <p>             Space, the final frontier. These are the voyages of the Starship             Enterprise. Its five-year mission: to explore strange new worlds, to             seek out new life and new civilizations, to boldly go where no man             has gone before.           </p>           <p>             As I stand out here in the wonders of the unknown at Hadley, I sort             of realize there’s a fundamental truth to our nature, Man must             explore, and this is exploration at its greatest.           </p>           <p>             <span>Placeholder text by </span>             <a href='http://spaceipsum.com'>Space Ipsum</a>             <span> Photographs by </span>             <a href='https://www.flickr.com/photos/nasacommons/'>               NASA on The Commons             </a>           </p>",
    "headerImageUrl": "https://images.pexels.com/photos/949587/pexels-photo-949587.jpeg?auto=compress&cs=tinysrgb&dpr=1&w=500",
    "datetime": "2019-12-20T17:28:01.440Z",
    "updatedAt": "2019-12-20T17:28:01.440Z",
    "__v": 0
}
```

### Additional Notes
- The model about the `comment` will be updated later when comments are implemented
- After this feature:
    - I will try implementing creation, update, deletion of post using CKeditor (rich text-editor that I plan to use)
    - Search feature (by content) with filter (date, type, category)

### CODE REVIEW

**Comment by frinzekt** [comment]
Change has been made on a missing comma in `package.json` file which aroused from a partial push of changes ... (so that's why there was a green and red change on that line even though they looked very much the same)

**Comment by noobling** [comment]
Yes you should commit the yarn.lock its what other devs will use to emulate your environment

File: backend/controller/post.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  })
+}
+
+module.exports = {
+  getPost,
+  getPagePost,
+  getSpecificPost,
+  getTitleContain,
+  addPost,
+  addComment,
```
Comment (noobling) [inline backend/controller/post.js]: This will still break because `addComment` does not exist. I don't understand how you dev on a broken application?

**Comment by frinzekt** [comment]
Thanks for checking David. I did learn a lot doing some backend stuff, I will remove the unnecessary comments and console log.

At the moment, I am still working on creating a post, I do have a question. When do you create a pull request? From what I know, it's when you complete a specific feature in this case... I completed the backend API for post and I just did some **initial connection (R)** (i think this conflicts with the title of my pull request).

I will be free at the end of the holidays, but I would do some stuff, and will callout in discord, if I do have specific questions. Thanks again David :)

**Comment by noobling** [comment]
The reason why this pr isn't a complete feature is there is an api but its not being used. Merging it in won't really do anything. More importantly its not tested. You can solve this by connecting it to the frontend and testing it.

A pull request should in most cases be a complete feature that does not break anything. It depends on what the team decides but generally, in a professional project these are the must-haves of a pr. I don't enforce most of these because I know you all are still learning. 

* code must conform to existing patterns and styles in code base
* must be linted and formatted correctly
* adequate tests are written
  *  unit tests for backend
  *  at least one e2e test
  *  tests for frontend where it makes sense e.g. testing api calls
* code must obviously be tested even if its just manual
* must be written in a maintainable way. The obvious things are no large components, hard coding things that shouldn't, magic numbers, not using dependency injection when accessing the connection to db and more specific things to frameworks e.g. for React a component should either only be fully controlled or fully uncontrolled https://reactjs.org/docs/uncontrolled-components.html
* passes security checks

Hopefully this gives you some perspective on just how deep our profession is. If you want to learn more heres a bigger list https://github.com/noobling/david-learns/issues/29

**Comment by noobling** [comment]
@frinzekt make sure to at me or request review when you finish with those changes so i can get notified

**Comment by frinzekt** [comment]
I think I am done

I will implement other feature in the next pull request such as:
- loader
- file upload
- some post settings ??

a

**Comment by noobling** [comment]
@innocentmm @JKKim98 can you guys run it and see if it works my mongo is throwing auth errorr -_-  
image

**Comment by innocentmm** [comment]
I’ll test it tomorrow cheers.

**Comment by noobling** [comment]
@innocentmm @JKKim98  have you guys had a chance too look at this?

**Comment by frinzekt** [comment]
I seem to be having new issues with starting nodemon. I believe it is coming from the [commit #656d0a4](https://github.com/codersforcauses/home2health/pull/50/commits/656d0a46e248e46b55d238636c55c0a2e304185d)

> App listening on port 4000
Port 4000 is already in use

wherein I believe the cause is `.listen()` method is called more than once - from `./bin/www` and `app.js` as I've [read ](https://stackoverflow.com/questions/39322089/node-js-port-3000-already-in-use-but-it-actually-isnt)

**Pull request merged**